provisioner/terraform/resources.go

```diff
@@ -120,17 +120,86 @@ func ConvertResources(module *tfjson.StateModule, rawGraph string) ([]*proto.Res
 		}
 	}
 
+	type appAttributes struct {
+		AgentID      string `mapstructure:"agent_id"`
+		Name         string `mapstructure:"name"`
+		Icon         string `mapstructure:"icon"`
+		URL          string `mapstructure:"url"`
+		Command      string `mapstructure:"command"`
+		RelativePath bool   `mapstructure:"relative_path"`
+	}
+	// Associate Apps with agents.
+	for _, resource := range tfResources {
+		if resource.Type != "coder_app" {
+			continue
+		}
+		var attrs appAttributes
+		err = mapstructure.Decode(resource.AttributeValues, &attrs)
+		if err != nil {
+			return nil, xerrors.Errorf("decode app attributes: %w", err)
+		}
+		if attrs.Name == "" {
+			// Default to the resource name if none is set!
+			attrs.Name = resource.Name
+		}
+		for _, agent := range agents {
+			if agent.Id != attrs.AgentID {
+				continue
+			}
+			agent.Apps = append(agent.Apps, &proto.App{
+				Name:         attrs.Name,
+				Command:      attrs.Command,
+				Url:          attrs.URL,
+				Icon:         attrs.Icon,
+				RelativePath: attrs.RelativePath,
+			})
+		}
+	}
+
 	for _, resource := range tfResources {
 		if resource.Mode == tfjson.DataResourceMode {
 			continue
 		}
 		if resource.Type == "coder_agent" || resource.Type == "coder_agent_instance" || resource.Type == "coder_app" {
 			continue
 		}
+		agents := findAgents(resourceDependencies, agents, convertAddressToLabel(resource.Address))
+		for _, agent := range agents {
+			// Didn't use instance identity.
+			if agent.GetToken() != "" {
+				continue
+			}
+
+			// These resource types are for automatically associating an instance ID
+			// with an agent for authentication.
+			key, isValid := map[string]string{
+				"google_compute_instance":         "instance_id",
+				"aws_instance":                    "id",
+				"azurerm_linux_virtual_machine":   "id",
+				"azurerm_windows_virtual_machine": "id",
+			}[resource.Type]
```

Comment:
So I'm still pretty noobish when it comes to Terraform, but it seems like we're automatically mapping these resource types for _all_ agents. So for instance if there's both a `google_compute_instance`, and an `aws_instance`, one would overwrite the other?

I'm sure the above wouldn't be a common use case, but I imagine it would be possible to create a semi-complex Terraform template with multiple providers depending on certain selections? And I imagine this would be a use-case for having multiple agents, are there others?

If my questions make no sense, feel free to disregard 😄

Reply:
> If my questions make no sense, feel free to disregard :smile:

All questions I'm happy to answer!

> So for instance if there's both a google_compute_instance, and an aws_instance, one would overwrite the other?

There's a one -> many mapping of resource to agent. If a resource isn't using token authentication, we assume it's using the zero-trust providers we support: `*-instance-identity` under the `auth` parameter on an agent. An instance ID can be manually associated with the `coder_agent_instance` resource, but we automatically associate it with one of the listed resource types. There isn't an explicit reason we need to do this, and it might end up adding more confusion, so it's not impossible we remove it at some point.

Reply:
Ok, thanks for clarifying @kylecarbs! I don't think we need to make any changes to it at this point, but good to be aware of it.